### PR TITLE
CA-242028: Remove exauth pbis cache mechanism.

### PR DESCRIPTION
**Pull request by xiewei20082008**
This is a cache-related issue. Now we have two kinds of cache mechanism in PBIS:
- xapi code in `extauth_plugin_ADpbis.ml`
- PBIS native cache mechanism `/opt/pbis/bin/config --detail CacheEntryExpiry`
It is not necessary to have two cache mechanism at the same time. 

Dev test passed - with following three changes:
- remove cache mechanism in `extauth_plugin_ADpbis.ml`
- set `pbis native cache expiration` to 0
- update the subject information in DB once a user is logging in XS successfully.

### CODE REVIEW

**Comment by xiewei20082008** [comment]
Hi @jonludlam. Could you please help have a look at this PR? Thanks a lot.

File: ocaml/xapi/extauth_plugin_ADpbis.ml [hunk 1]
```diff
@@ -115,7 +115,7 @@ struct
     pbis_config "CreateHomeDir" "false";
     pbis_config "SyncSystemTime" "false";
     pbis_config "LdapSignAndSeal" "true";
-    pbis_config "CacheEntryExpiry" "300";
+    pbis_config "CacheEntryExpiry" "0";
```
Comment (mseri) [inline ocaml/xapi/extauth_plugin_ADpbis.ml]: Why is this necessary?

**Comment by mseri** [comment]
I have a couple of questions:

- The other invocations of `pbis_common` and the equivalent with password are not behind a mutex? I think the original ones where because we needed to guarantee the consistency of our cache, but it looks to me now that either the pbis commands are already doing that for us and we don't need the mutex at all, or we need to protect everything behind such mutex

- Can you please include your PR comment in the commit message?

- I don't yet have the context to understand when `Xapi_subject.update_all_subjects` should be called, can you explain the rationale behind the two new calls please?

**Comment by xiewei20082008** [comment]
@mseri Thanks for your review! For your questions:

1. mutex

I also think the mutex is only for the consistency for the original cache. I just kept that in case of other considerations. 

I think we can remove the mutex. If so, I will do more tests to confirm.

2. commit message

I didn't get what you mean. I thought the commit message is the same with the PR comments. Isn't it?

3. Xapi_subject.update_all_subjects

This issue is that the XC shows fault info of AD users (in xapi db) when we reconnect the XC to the XS, So we have to update the AD users info in  xapi db when a new session occurs.

Please let me know if I didn't make it clear.

File: ocaml/xapi/extauth_plugin_ADpbis.ml (hunk 1 above)
Comment (xiewei20082008) [inline ocaml/xapi/extauth_plugin_ADpbis.ml]: If we don't do this. The CA issue will still exist, since a cache data sometimes is also a wrong data. Moreover, the pbis command is not frequently being invoked. It seems there is no much need to have a cache here.

**Comment by mseri** [comment]
Thanks!

1. I think we should remove the mutex there.

2. The commit message is empty, only the title matches. Please add more context to the commit message, I think the comment you added to this PR would be fine.

3. I see, are those the only places where this issue can appear?

**Comment by jonludlam** [comment]
This cache was added because the performance was considered bad - I'd quite like to go and check the original tests that indicated this before I'll be convinced that this is a good idea. @mg12 would probably have an opinion.

**Comment by jonludlam** [comment]
CP-12169 and CA-159776 were the tickets.

**Comment by mg12** [comment]
@xiewei20082008 I think we should remeasure the login times to make sure that removing the cache will not cause a regression. These results should be part of the pull request and commit messages. I've added more details in CA-242028.

**Comment by xiewei20082008** [comment]
Thanks @jonludlam @mg12 . I will do the test to check the performance.

**Comment by mseri** [comment]
Any news on the performance impact of this change?

**Comment by xiewei20082008** [comment]
Hi @mseri . I didn't have much time on this ticket recently. I am still working on it. Here is the progress:
-  The script for setting up the performance test did not work due to the upgrading of pyad lib. I've make some slight changes to get it work.
- The performance test environment (1000000 users) have been set up successfully.

**Comment by thomassa** [comment]
I have just added the "needs updating" label so that a reviewer looking at the list of open pull-requests can see at a glance that there is no need to examine this one.
The code might turn out not to need changes, but this seems the most appropriate of the existing labels because the next action (performance testing) is to be taken by the developer not the reviewer.

**Comment by Zhengchai** [comment]
Weixie. we reviewed this on S28 Sprint Review, team suggested we have a local internal discussion and review first on this one please.

**Comment by xiewei20082008** [comment]
Hi @thomassa . Sorry for the inconvenience that all you guys examined the PR repetitively. I know that. It is because we are attending the new-hiring and XAXD training for entire two weeks! so the performance test is stuck. It is great to have a "needs updating" label. Thanks!

I will immediately back to the PR once I get effort.

**Comment by xiewei20082008** [comment]
Hi @mseri ,I updated the PR to remove the `exauth cache`, while keeping the native PBIS cache. I finished the performance test [https://info.citrite.net/display/eng/Deeper+performance+test+for+PBIS], which shows `exauth cache` do not have any effects at all. The detail info can be found in [https://issues.citrite.net/browse/CA-242028]. Could you help review the code again, please? Thanks!

**Comment by xiewei20082008** [comment]
As Jonathan Davies commented, There are two sequences that generate performance measurements: `dundeetc8412.seq` and `dundeetc8412WE.seq`. I will have the two XenRT case tested. 

Status: needs updating.

**Comment by mseri** [comment]
If there is no action on this pull request can we close it for the moment and reopen it once its status is clearer?

**Comment by xiewei20082008** [comment]
Hi @mseri . I've finished the XenRT test which shows there is no performance regression - waiting for Marcus to confirm. From my perspective, this PR can be merged to remove the xauth cache.

**Comment by xiewei20082008** [comment]
Hi @mg12 , could you help review the PR?  Thanks.

**Comment by mg12** [comment]
@xiewei20082008 the code involving removing the extauth cache seems ok, the only questions I have is related to some details of the measurements done, which I would like to understand better. I've added the measurement questions to CA-242028.

**Comment by xiewei20082008** [comment]
As talked with Marcus. Closing the PR.
